### ISSUE TITLE
Failures new nnbd VM builders on ia32

### ISSUE THREAD

**Issue by a-siva** [dart-lang/sdk#43613]
There are new test failures on [[infra] Add additional nnbd VM builders](https://github.com/dart-lang/sdk/commit/c3d37c1ad6d937a9cef1c1e4579f3ca4090fab72).

```
vm/dart/type_casts_with_null_safety_autodetection_test
vm/dart/appjit_determinism_test
vm/dart/appjit_cha_deopt_test
vm/dart/appjit_load_static_licm_test
vm/dart/product_aot_kernel_test
lib/isolate/nnbd_spawn_autodetect_1_test
lib/isolate/detect_nullsafety_1_test
lib/isolate/nnbd_spawn_autodetect_2_test
lib/isolate/detect_nullsafety_2_test
```

**Comment by a-siva** [comment]
Additional Tests failing on simarm 
```
standalone/http_launch_test/0
standalone/http_launch_test/1
standalone/http_launch_test/2
standalone/http_launch_test/3
```

**Comment by alexmarkov** [comment]
Triaging `vm/dart` failures:

`vm/dart/type_casts_with_null_safety_autodetection_test` is just slow in debug mode as it runs with `--optimization-counter-threshold=10` and on IA32 we don't use kernel service snapshot.

Failures
```
vm/dart/appjit_determinism_test
vm/dart/appjit_cha_deopt_test
vm/dart/appjit_load_static_licm_test
```
are extracted to a separate issue https://github.com/dart-lang/sdk/issues/43626.

`vm/dart/product_aot_kernel_test` fails in strong mode as it depends on package:kernel, package:vm which are not migrated yet.  The same compile-time error happens on x64 and tracked in https://github.com/dart-lang/sdk/issues/42146.

**Comment by alexmarkov** [comment]
`lib/isolate/nnbd_spawn_autodetect_1_test` and `lib/isolate/detect_nullsafety_1_test` are also failing due to https://github.com/dart-lang/sdk/issues/43626.

`lib/isolate/nnbd_spawn_autodetect_2_test` and `lib/isolate/detect_nullsafety_2_test` are timing out. These tests also create app-jit snapshots, but they dodged https://github.com/dart-lang/sdk/issues/43626 and were able to create snapshots. On IA32 those snapshots have `Snapshot::kFull` kind and do not contain the code. When running from those snapshots, null safety mode auto-detection doesn't query mode from the snapshot as snapshot doesn't contain code:

https://github.com/dart-lang/sdk/blob/f54c0a3b9bfe0d2cde5d84625c5bd3b046ed6f79/runtime/vm/dart.cc#L780-L788

After that, auto-detection logic falls back to kernel service for detection and passes snapshot as the main script URI, and CFE hangs while trying to parse snapshot file as a main `.dart` file.

It seems like we need to embed null safety mode into `Snapshot::kFull` application snapshots on IA32, and detect mode from such snapshots even if they do not contain code. We shouldn't really try to call kernel service if running from snapshot.

**Comment by a-siva** [comment]
Another ia32 specific failure ` co19/LibTest/isolate/Isolate/spawn_A04_t04`

**Comment by a-siva** [comment]
Another ia32 specific failure `co19/Language/Libraries_and_Scripts/Scripts/top_level_main_t01`

**Comment by alexmarkov** [comment]
`standalone/regress_42092_test` times out on dartk-weak-asserts-linux-release-ia32 (maybe flaky). Log: https://dart-ci.appspot.com/log/vm-kernel-nnbd-linux-release-ia32/dartk-weak-asserts-linux-release-ia32/3/standalone/regress_42092_test

**Comment by alexmarkov** [comment]
`co19/LibTest/isolate/Isolate/spawn_A04_t04` is flaky, similarly to `co19_2/LibTest/isolate/Isolate/spawn_A04_t04` and not specific to ia32 nor nnbd (fails on all architectures and both legacy and nnbd bots).

`co19/Language/Libraries_and_Scripts/Scripts/top_level_main_t01` fails on all architectures, corresponding legacy test `co19_2/Language/Libraries_and_Scripts/Scripts/top_level_main_t01` fails similarly.

**Comment by alexmarkov** [comment]
`standalone/regress_42092_test` is also flaky and failing on all architectures and both nnbd and legacy bots.

**Comment by alexmarkov** [comment]
Summary:

`vm/dart/type_casts_with_null_safety_autodetection_test` is addressed in 34e8b4229efec853ae171dc1569c03b1d053c58f.

The tests
```
vm/dart/appjit_determinism_test
vm/dart/appjit_cha_deopt_test
vm/dart/appjit_load_static_licm_test
lib/isolate/nnbd_spawn_autodetect_1_test
lib/isolate/detect_nullsafety_1_test
lib/isolate/nnbd_spawn_autodetect_2_test
lib/isolate/detect_nullsafety_2_test
```
are fixed in d77fff73075426306d2c3d77ccf61afd3eba9ab1.

`vm/dart/product_aot_kernel_test` is not specific to ia32 and tracked in https://github.com/dart-lang/sdk/issues/42146.

Other failures mentioned above are not specific to NNBD, and not specific to ia32.

**Comment by sstrickl** [comment]
It looks like vm/dart/type_casts_with_null_safety_autodetection_test has still been at least flakily timing out since Oct 28 in release mode, and finally went red today ([log](https://dart-ci.appspot.com/log/vm-kernel-nnbd-win-release-ia32/dartk-weak-asserts-win-release-ia32/473/vm/dart/type_casts_with_null_safety_autodetection_test)). Will also mark it Slow in release mode.
